File: ThisIsCodingTest/이것이코딩테스트다_18장_탑승구.py
class DisjointSet:
    def __init__(self, n):
        self.p = [0] * (n + 1)

        for i in range(1, n + 1):
            self.make_set(i)

    def make_set(self, x):
        self.p[x] = x

    def find_set(self, x):
        if self.p[x] != x:
            self.p[x] = self.find_set(self.p[x])
        return self.p[x]

    def union_set(self, x, y):
        px = self.find_set(x)
        py = self.find_set(y)
        self.p[px] = py


G = int(input())
P = int(input())

gates = []
for _ in range(P):
    g = int(input())
    gates.append(g)

disjointSet = DisjointSet(G)
answer = 0

for gate in gates:
    g = disjointSet.find_set(gate)

    if g == 0:
        break

    disjointSet.union_set(g, g - 1)
    answer += 1

print(answer)

# 4
# 3
# 4
# 1
# 1
# => 2

File: this-is-coding-test/이것이코딩테스트다_18장_탑승구.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {
    public static void main(String[] args) throws IOException {
        Main main = new Main();
        System.out.println(main.solution());
    }

    public int solution() throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int G = Integer.parseInt(br.readLine());
        int P = Integer.parseInt(br.readLine());

        DisjointSet disjointSet = new DisjointSet(G);
        int cnt = 0;

        for (int i = 0; i < P; i++) {
            int g = Integer.parseInt(br.readLine());
            int f = disjointSet.findSet(g);

            // 루트 노드가 0인 경우 도킹을 할 수 있는 탑승구가 없다
            if (f == 0) {
                break;
            }

            disjointSet.unionSet(f - 1, f);
            cnt += 1;
        }

        return cnt;
    }
}

class DisjointSet {
    int n;
    int[] p;

    DisjointSet(int n) {
        this.n = n;
        p = new int[n + 1];
        for (int i = 0; i < n + 1; i++) {
            makeSet(i);
        }
    }

    void makeSet(int x) {
        p[x] = x;
    }

    int findSet(int x) {
        if (p[x] != x) {
            p[x] = findSet(p[x]);
        }
        return p[x];
    }

    void unionSet(int x, int y) {
        int px = findSet(x);
        int py = findSet(y);

        if (p[px] >= p[py]) {
            p[px] = py;
        } else {
            p[py] = px;
        }
    }
}
